realm/realm-library/src/main/cpp/jni_util/log.cpp

```diff
@@ -0,0 +1,193 @@
+    jmethodID m_log_method;
+
+    inline JNIEnv* get_current_env() noexcept
+    {
+        JNIEnv *env;
+        if (m_jvm->GetEnv((void **)&env, JNI_VERSION_1_6) != JNI_OK) {
+            m_jvm->AttachCurrentThread(&env, nullptr); // Should never fail
+        }
+        return env;
+    }
+};
+
+JniLogger::JniLogger()
+    :m_is_java_logger(false)
+{
+}
+
+JniLogger::JniLogger(bool is_java_logger)
+    :m_is_java_logger(is_java_logger)
+{
+}
+
+JavaLogger::JavaLogger(JNIEnv* env, jobject java_logger)
+    :JniLogger(true)
+{
+    jint ret = env->GetJavaVM(&m_jvm);
+    if (ret != 0) {
+        throw std::runtime_error(util::format("Failed to get Java vm. Error: %d", ret));
+    }
+    m_java_logger = env->NewGlobalRef(java_logger);
+    jclass cls = env->GetObjectClass(m_java_logger);
+    m_log_method = env->GetMethodID(cls, "log", "(ILjava/lang/String;Ljava/lang/Throwable;Ljava/lang/String;)V");
+}
+
+JavaLogger::~JavaLogger()
+{
+    get_current_env()->DeleteGlobalRef(m_java_logger);
+}
+
+void JavaLogger::log(Log::Level level, const char* tag, jthrowable throwable, const char*,
+        const char* message)
+{
+    JNIEnv *env = get_current_env();
+
+    // NOTE: If there are Java exception has been thrown in native code, the below call will trigger an JNI exception
+    // "JNI called with pending exception". This is something should be avoided when printing log in JNI -- Always
+    // print log before calling env->ThrowNew. Doing env->ExceptionCheck() here creates overhead for normal cases.
+    env->CallVoidMethod(m_java_logger, m_log_method, level, env->NewStringUTF(tag),
+            throwable, env->NewStringUTF(message));
+}
+
+bool JavaLogger::is_same_object(JNIEnv* env, jobject java_logger)
+{
+    return env->IsSameObject(m_java_logger, java_logger);
+}
+
+Log::Log()
+    : m_loggers()
+{
+    add_logger(get_default_logger());
+}
+
+Log& Log::shared()
+{
+    static Log log;
+    return log;
+}
+
+void Log::add_java_logger(JNIEnv* env, const jobject java_logger)
+{
+    std::shared_ptr<JniLogger> logger = std::make_shared<JavaLogger>(env, java_logger);
+    add_logger(logger);
+}
+
+void Log::remove_java_logger(JNIEnv* env, const jobject java_logger)
+{
+    std::lock_guard<std::mutex> lock(m_mutex);
+    for (auto it = m_loggers.begin(); it != m_loggers.end(); ++it) {
+        if ((*it)->m_is_java_logger &&
+                std::static_pointer_cast<JavaLogger>(*it)->is_same_object(env, java_logger)) {
+            m_loggers.erase(it);
+            break;
+        }
+    }
+}
+
+void Log::add_logger(std::shared_ptr<JniLogger> logger)
+{
+    std::lock_guard<std::mutex> lock(m_mutex);
+    m_loggers.push_back(logger);
+}
+
+void Log::remove_logger(std::shared_ptr<JniLogger> logger)
+{
+    std::lock_guard<std::mutex> lock(m_mutex);
+    for (auto it = m_loggers.begin(); it != m_loggers.end(); ++it) {
+        if (*it == logger) {
+            m_loggers.erase(it);
+        }
+    }
+}
+
+void Log::clear_loggers()
```

Comment:
That sounds like a test problem that shouldn't really leak into the API?  What about just re-adding the logger after the test?